### PR TITLE
[WAF] Add resource huaweicloud_waf_dedicated_instance to managing WAF dedicated instances.

### LINKED ISSUE #1355: [WAF] Add data sources and resources for dedicated engine.
### Current Terraform Version

```
1.27.0
```

This is an internal requirement as follows:
- Add the data source to query WAF policies, source name is `huaweicloud_waf_policies`.
- Manage domain names in Premium-waf, resource name is `huaweicloud_waf_dedicated_domain`.
  Functions include: creating a domain names, modifying protection status, modifying protection mode, and deleting domain names.
- Manage engine in Premium-waf, resource name is `huaweicloud_waf_dedicated_engine`.
  Functions include: creating a dedicated engine,deleting dedicated,querying dedicated engine by id.
- We also needs to query the dedicated engine infomation, the data source name is `huaweicloud_waf_dedicated_engines` and `huaweicloud_waf_dedicated_engine`.

### PR DESCRIPTION
**Pull request by chengxiangdong**
**What this PR does / why we need it**:

- Add the resource to manage WAF dedicated instance, resource name is `huaweicloud_waf_dedicated_instance`.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1355

**Special notes for your reviewer**:

NONE

**Release note**:

```release-note
NONE
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/waf' TESTARGS='-run TestAccWafDedicatedInstanceV1_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/waf -v -run TestAccWafDedicatedInstanceV1_basic -timeout 360m -parallel 4
=== RUN   TestAccWafDedicatedInstanceV1_basic
--- PASS: TestAccWafDedicatedInstanceV1_basic (397.64s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/waf       397.766s

## Test for TestAccServiceEndpoints_Security of endpoints_test.go
make testacc TEST='./huaweicloud' TESTARGS='-run TestAccServiceEndpoints_Security'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run TestAccServiceEndpoints_Security -timeout 360m -parallel 4
=== RUN   TestAccServiceEndpoints_Security
2021/08/25 20:35:04 [DEBUG] customer endpoints: map[]
...
...
    endpoints_test.go:280: anti-ddos endpoint:   https://antiddos.cn-north-4.myhuaweicloud.com/v1/0970dd7a1300f5672ff2c003c60ae115/
    endpoints_test.go:292: KMS endpoint:         https://kms.cn-north-4.myhuaweicloud.com/v1.0/
    endpoints_test.go:304: SCM endpoint:         https://scm.cn-north-4.myhuaweicloud.com/v3/
    endpoints_test.go:316: WAF endpoint:         https://waf.cn-north-4.myhuaweicloud.com/v1/0970dd7a1300f5672ff2c003c60ae115/waf/
    endpoints_test.go:328: WAF dedicated endpoint:       https://waf.cn-north-4.myhuaweicloud.com/v1/0970dd7a1300f5672ff2c003c60ae115/premium-waf/
--- PASS: TestAccServiceEndpoints_Security (1.82s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       1.971s

```

### CODE REVIEW

File: examples/waf/waf-dedicated-instance/main.tf [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    huaweicloud_networking_secgroup_v2.secgroup.id
+  ]
+}
+
+resource "huaweicloud_waf_policy" "policy_1" {
+  name  = var.waf_policy_name
+  level = 1
+
+  depends_on = [huaweicloud_waf_dedicated_instance.instance_1] # Make sure that a dedicated instance has been created.
+}
```
Comment (ShiChangkuo) [inline examples/waf/waf-dedicated-instance/main.tf]: No new line at end of file

File: huaweicloud/config/endpoints.go [hunk 2]
```diff
@@ -263,6 +263,11 @@ var allServiceCatalog = map[string]ServiceCatalog{
 		Version:      "v1",
 		ResourceBase: "waf",
 	},
+	"waf-dedicated": {
```
Comment (ShiChangkuo) [inline huaweicloud/config/endpoints.go:266]: please add acceptance test for the client

File: docs/resources/waf_dedicated_instance.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+variable az_name {}
+variable cpu_flavor_id {}
+variable vpc_id {}
+variable subnet_id {}
+variable security_group_id {}
+
+resource "huaweicloud_waf_dedicated_instance" "instance_1" {
+  name               = "instance_1"
+  available_zone     = var.az_name
+  cpu_architecture   = "x86"
```
Comment (ShiChangkuo) [inline docs/resources/waf_dedicated_instance.md]: user can ignore `cpu_architecture`, as it defaults to x86

File: docs/resources/waf_dedicated_instance.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+* `available_zone` - (Required, String, ForceNew) The available zone names for the dedicated instances. It can be obtained through 
+  this data source `huaweicloud_availability_zones`. Changing this will create a new instance.
+
+* `specification_code` - (Required, String, ForceNew) The specification code of instance. 
+  Different specifications have different throughput. Changing this will create a new instance. Values are:
+  * `waf.instance.professional` - The professional edition, throughput: 100 Mbit/s; QPS: 2,000 (Reference only).
+  * `waf.instance.enterprise` - The enterprise edition, throughput: 500 Mbit/s; QPS: 10,000 (Reference only).
+  
+* `cpu_architecture` - (Optional, String, ForceNew) The ECS cpu architecture of instance, e.g. : `x86`.
```
Comment (ShiChangkuo) [inline docs/resources/waf_dedicated_instance.md]: suggest put the optional arguments after required ones.

File: docs/resources/waf_dedicated_instance.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  Changing this will create a new instance.
+
+* `cpu_flavor` - (Required, String, ForceNew) The ECS specification, the flavors of the ECS can be obtained through this
```
Comment (ShiChangkuo) [inline docs/resources/waf_dedicated_instance.md]: can the argument named as `ecs_flavor`  or `compute_flavor` ?

File: huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+		PreCheck:     func() { acceptance.TestAccPreCheck(t) },
+		Providers:    acceptance.TestAccProviders,
+		CheckDestroy: testAccCheckWafDedicatedInstanceV1Destroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccWafDedicatedInstanceV1_conf(name),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckWafDedicatedInstanceV1Exists(resourceName, &instance),
+					resource.TestCheckResourceAttr(resourceName, "name", name),
+					resource.TestCheckResourceAttr(resourceName, "available_zone", "cn-north-4b"),
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go]: hard coding is not recommended.

File: huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+					resource.TestCheckResourceAttrSet(resourceName, "subnet_id"),
+					resource.TestCheckResourceAttrSet(resourceName, "vpc_id"),
+					resource.TestCheckResourceAttrSet(resourceName, "cpu_flavor"),
+				),
+			},
+			{
+				ResourceName:            resourceName,
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{},
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go]: you can delete this line as the array is empty.

File: huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+resource "huaweicloud_vpc_subnet" "vpc_subnet_1" {
+  name          = "%s_waf"
+  cidr          = "192.168.0.0/24"
+  gateway_ip    = "192.168.0.1"
+  primary_dns   = "100.125.1.250"
+  secondary_dns = "100.125.129.250"
+  vpc_id        = huaweicloud_vpc.vpc_1.id
+}
+
+resource "huaweicloud_networking_secgroup_v2" "secgroup" {
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go]: please use **huaweicloud_networking_secgroup** instead.

File: huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+  name = "%s_waf"
+  cidr = "192.168.0.0/24"
+}
+
+resource "huaweicloud_vpc_subnet" "vpc_subnet_1" {
+  name          = "%s_waf"
+  cidr          = "192.168.0.0/24"
+  gateway_ip    = "192.168.0.1"
+  primary_dns   = "100.125.1.250"
+  secondary_dns = "100.125.129.250"
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go]: you can use the default dns as the region is not fixed to cn-north-4.

**Review by ShiChangkuo — CHANGES_REQUESTED** [review]

File: docs/resources/waf_dedicated_instance.md (hunk 3 above)
Comment (chengxiangdong) [reply docs/resources/waf_dedicated_instance.md]: removed

File: docs/resources/waf_dedicated_instance.md (hunk 5 above)
Comment (chengxiangdong) [reply docs/resources/waf_dedicated_instance.md]: `ecs_flavor` is better.

File: huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go (hunk 6 above)
Comment (chengxiangdong) [reply huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go]: got it.

File: huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go (hunk 7 above)
Comment (chengxiangdong) [reply huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go]: deleted

File: huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go (hunk 9 above)
Comment (chengxiangdong) [reply huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go]: yes, it should use default dns

File: huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go (hunk 8 above)
Comment (chengxiangdong) [reply huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_dedicated_instance_test.go]: changed

File: huaweicloud/endpoints_test.go [hunk 10]
```diff
@@ -302,6 +302,30 @@ func TestAccServiceEndpoints_Security(t *testing.T) {
 		t.Fatalf("SCM endpoint: expected %s but got %s", green(expectedURL), yellow(actualURL))
 	}
 	t.Logf("SCM endpoint:\t %s", actualURL)
+
+	// test the endpoint of WAF service
+	serviceClient, err = config.WafV1Client(HW_REGION_NAME)
+	if err != nil {
+		t.Fatalf("Error creating HuaweiCloud WAF client: %s", err)
+	}
+	expectedURL = fmt.Sprintf("https://waf.%s.%s/v3/", HW_REGION_NAME, config.Cloud)
```
Comment (ShiChangkuo) [inline huaweicloud/endpoints_test.go]: v1 or v3?

Comment (chengxiangdong) [reply huaweicloud/endpoints_test.go]: sorry, it is v1.

File: huaweicloud/config/endpoints.go (hunk 2 above)
Comment (chengxiangdong) [reply huaweicloud/config/endpoints.go:266]: added

**Pull request merged**